### PR TITLE
Obstacle flow and anti-bounce-back outlet

### PR DESCRIPTION
**Pull request by MartinKliemank**
Cleaned up all (or most of) the things, made outlet boundary work on any border of the domain (hopefully)

There is probably still room for improvement / compacting things

### CODE REVIEW

File: lettuce/boundary.py [hunk 1]
```diff
@@ -37,3 +38,63 @@ def __call__(self, f):
...
+
+    def __call__(self, f):
+        u = self.lattice.u(f)
+        u_w = u[[slice(None)] + self.index] + 0.5 * (u[[slice(None)] + self.index] - u[[slice(None)] + self.neighbour])
+        for i in self.velocities:
+            f[[self.lattice.stencil.opposite[i]] + self.index] = - f[[i] + self.index] + self.lattice.stencil.w[i] * self.lattice.rho(f)[[0] + self.index] * \
+                     (2 + torch.einsum(self.dims, torch.tensor(self.lattice.stencil.e[i], device=f.device, dtype=f.dtype), u_w) ** 2 / self.lattice.stencil.cs ** 4 - (torch.norm(u_w, dim=0) / self.lattice.stencil.cs)**2)
+        return f
+
+#    def __call__(self, f):
```
Comment (dominikwilde) [inline lettuce/boundary.py:83]: As the generic BB scheme works, please remove the unneeded passages from the code.

File: lettuce/boundary.py [hunk 2]
```diff
@@ -37,3 +38,63 @@ def __call__(self, f):
...
+
+class AntiBounceBackOutlet:
+    """Allows distributions to leave domain unobstructed through this boundary.
+        Based on equations from page 195 of "The lattice Boltzmann method" (2016 by Krüger et al.)
+        Give the planned side of the boundary as array [1, 0, 0] for positive x-direction in 3D; [1, 0] for the sam in 2D
+        [-1, 0, 0] is negative x-direction ect. [x, y, z]; only one entry nonzero
+        """
+
+    def __init__(self, lattice, direction):
+        direction = np.array(direction)
```
Comment (dominikwilde) [inline lettuce/boundary.py:50]: Assert if direction has the correct shape

File: lettuce/flows/obstacle.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import numpy as np
+import torch
+
+from lettuce.unit import UnitConversion
+from lettuce.boundary import EquilibriumBoundaryPU, BounceBackBoundary, AntiBounceBackOutlet
+
+
+class Obstacle2D(object):
+
+    def __init__(self, resolution_x, resolution_y, reynolds_number, mach_number, lattice, char_length_lu=64):
```
Comment (dominikwilde) [inline lettuce/flows/obstacle.py:10]: Remove the default value 64

File: lettuce/boundary.py [hunk 4]
```diff
@@ -37,3 +38,63 @@ def __call__(self, f):
...
+            self.dims = 'c, xy -> '.replace('xy'[(np.where(abs(direction)==1))[0][0]], 'c') + 'xy'.replace('xy'[(np.where(abs(direction)==1))[0][0]], '')
+        if len(direction) == 1:
+            self.dims = 'c, x -> x'
+
+    def __call__(self, f):
+        u = self.lattice.u(f)
+        u_w = u[[slice(None)] + self.index] + 0.5 * (u[[slice(None)] + self.index] - u[[slice(None)] + self.neighbour])
+        for i in self.velocities:
+            f[[self.lattice.stencil.opposite[i]] + self.index] = - f[[i] + self.index] + self.lattice.stencil.w[i] * self.lattice.rho(f)[[0] + self.index] * \
+                     (2 + torch.einsum(self.dims, torch.tensor(self.lattice.stencil.e[i], device=f.device, dtype=f.dtype), u_w) ** 2 / self.lattice.stencil.cs ** 4 - (torch.norm(u_w, dim=0) / self.lattice.stencil.cs)**2)
```
Comment (dominikwilde) [inline lettuce/boundary.py:80]: use lattice.e instead of lattice.stencil.e

(should we remove stencil as a class member of lattice?)

File: lettuce/boundary.py [hunk 5]
```diff
@@ -37,3 +38,63 @@ def __call__(self, f):
...
+
+    def __init__(self, lattice, direction):
+        direction = np.array(direction)
+        self.lattice = lattice
+        #select velocities to be bounced (the ones pointing in "direction")
+        self.velocities = np.concatenate(np.argwhere(np.matmul(self.lattice.stencil.e, direction) == 1), axis=0)
+        #build indices of u/f that determine the side of the domain
+        self.index = []
+        self.neighbour = []
+        #self.index = np.where(direction == 0, [slice(None)], [slice((direction + 1)*-0.5)])
```
Comment (dominikwilde) [inline lettuce/boundary.py:57]: unneeded comment

File: lettuce/simulation.py [hunk 6]
```diff
@@ -50,7 +50,8 @@ def __init__(self, flow, lattice, collision, streaming):
         self.no_collision_mask = lattice.convert_to_tensor(self.no_collision_mask)
         for boundary in self.flow.boundaries:
             if boundary.__class__.__name__ == "BounceBackBoundary":
-                self.no_collision_mask = boundary.mask | self.no_collision_mask
+                self.no_collision_mask = boundary.mask.bool() | self.no_collision_mask.bool()
+                self.no_collision_mask = lattice.convert_to_tensor(self.no_collision_mask).byte()
```
Comment (dominikwilde) [inline lettuce/simulation.py:54]: Related to the type conversion issue in boundary.py. This one would be okay

File: tests/test_boundaries.py [hunk 7]
```diff
@@ -47,3 +47,28 @@ def test_equilibrium_boundary_pu(f_lattice):
     assert f.cpu().numpy() == pytest.approx(feq_field.cpu().numpy())
     #assert f.cpu().numpy() == pytest.approx(f_old.cpu().numpy())
 
+def test_anti_bounce_back_outlet(f_lattice):
```
Comment (dominikwilde) [inline tests/test_boundaries.py:50]: Please comment on what it does.

**Review by dominikwilde — COMMENTED** [review]
Great!
Please have a look at the errors in TravisCI and find my suggestions below.

File: lettuce/boundary.py [hunk 8]
```diff
@@ -14,7 +15,7 @@ def __init__(self, mask, lattice):
         self.lattice = lattice
 
     def __call__(self, f):
-        f = torch.where(self.mask, f[self.lattice.stencil.opposite], f)
+        f = torch.where(self.mask.byte(), f[self.lattice.stencil.opposite], f)
```
[the remark this answers is not in the document]
Comment (MartinKliemank) [reply lettuce/boundary.py:18]: moved it to the init of the bb_boundary... hope thats fine?

File: lettuce/boundary.py (hunk 2 above)
Comment (MartinKliemank) [reply lettuce/boundary.py:50]: cue totally over complicated assert checking for all the possible errors (the function only accepts one kind of input, so i feel this is justified)

File: lettuce/boundary.py (hunk 4 above)
Comment (MartinKliemank) [reply lettuce/boundary.py:80]: just noticed, lattice doesn't provide lattice.stencil.opposite, so that would have to be added in case of a removal

Comment (Olllom) [reply lettuce/boundary.py:80]: I strongly advocate for keeping stencil as an attribute of lattice

File: lettuce/boundary.py (hunk 2 above)
Comment (Olllom) [reply lettuce/boundary.py:50]: It's well documented, that is enough for me.

File: lettuce/boundary.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Olllom) [reply lettuce/boundary.py:18]: fine with me

File: lettuce/boundary.py (hunk 2 above)
Comment (MartinKliemank) [reply lettuce/boundary.py:54]: The assert is now pushed, so you can see if that is helpful or too much. I was hoping all the outputs would make it easy to fix errors?

File: tests/test_boundaries.py (hunk 7 above)
Comment (MartinKliemank) [reply tests/test_boundaries.py:50]: Are the comments enough? Its pretty rough, but I was unsure what else to explain here.

Comment (MartinKliemank) [reply tests/test_boundaries.py:50]: (I've added a reference to the book instead of "the source" now, please don't mind that part)

**Comment by MartinKliemank** [comment]
I think all surface level issues are fixed now? but a fundamental problem still remains: the boundary is only stable when opposite to the equilibrium boundary. When this is given having two inlets and outlets is no problem for example. However, using two anti-bounce-back outlets opposite each other causes instability to spread from them. I have not yet been able to figure out what is causing this issue.

File: lettuce/flows/obstacle.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    @property
+    def boundaries(self):
+        x, y = self.grid
+        return [EquilibriumBoundaryPU(np.abs(x) < 1e-6, self.units.lattice, self.units, np.array(
+            [self.units.characteristic_velocity_pu, self.units.characteristic_velocity_pu * 0.0])),
+                AntiBounceBackOutlet(self.units.lattice, [1, 0]),
+                BounceBackBoundary(self.mask, self.units.lattice)]
+
+
+class Obstacle3D(object):
```
Comment (Olllom) [inline lettuce/flows/obstacle.py:42]: Same comments as for 2D

File: lettuce/simulation.py (hunk 6 above)
Comment (Olllom) [inline lettuce/simulation.py:54]: Why do this with bool and byte?

**Review by Olllom — COMMENTED** [review]
Thanks a lot. This looks very good.
Please see details and questions below.

File: lettuce/flows/obstacle.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+            characteristic_velocity_pu=1
+        )
+        self.mask = None
+
+    def initialise_object(self, mask, lattice):
+        self.boundaries.append(BounceBackBoundary(mask, lattice))
+
+    def initial_solution(self, x):
+        return np.array([0 * x[0]], dtype=float), np.array(
+            [0 * x[0] + self.units.convert_velocity_to_lu(self.units.characteristic_velocity_pu), 0 * x[1]],
```
[the remark this answers is not in the document]
Comment (MartinKliemank) [reply lettuce/flows/obstacle.py:24]: Yes, that seems like an oversight, i think it was even being converted twice, to a false value.

File: lettuce/simulation.py (hunk 6 above)
Comment (MartinKliemank) [reply lettuce/simulation.py:54]: Apparently since pytorch 1.2 (lettuce seems to be on 1.4) pytorch also includes a bool tensor, which torch.where ect. can work with, so the byte()-conversion for unit8 is no longer necessary.
I've swapped out the uint8 related stuff against bool now, so we can save a bit of memory and make the function of the masks more obvious. Also the masks are now bool by default, so the logical operation does not need any conversion before hand.

Comment (MartinKliemank) [reply lettuce/simulation.py:54]: Apparently since pytorch 1.2 (lettuce seems to be on 1.4) pytorch also includes a bool tensor, which torch.where ect. can work with, so the byte()-conversion for unit8 is no longer necessary.
I've swapped out the uint8 related stuff against bool now, so we can save a bit of memory and make the function of the masks more obvious. Also the masks are now bool by default, so the logical operation does not need any conversion before hand.

File: tests/test_flows.py [hunk 11]
```diff
@@ -62,3 +63,23 @@ def test_divergence(stencil, dtype_device):
...
+    if stencil is D2Q9:
+        mask = np.zeros([20, 10])
+        mask[3:6, 3:6] = np.ones([3, 3]) > 0
+        flow = Obstacle2D(20, 10, 100, 0.1, lattice=lattice, char_length_lu=3)
+    if stencil is D3Q27:
+        mask = np.zeros([20, 10, 5])
+        mask[3:6, 3:6, :] = np.ones([3, 3, 5]) > 0
+        flow = Obstacle3D(20, 10, 5, 100, 0.1, lattice=lattice, char_length_lu=3)
+    collision = BGKCollision(lattice, tau=flow.units.relaxation_parameter_lu)
+    mask = torch.tensor(mask, dtype=torch.bool)
```
Comment (Olllom) [inline tests/test_flows.py:81]: Please make sure that this works with numpy masks. The basic philosophy here is that flows are defined in numpy while everything in the simulation uses torch.
